rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,643 @@
+                1 as symcrypt_sys::UINT32,
+                primes_ptr,
+                primes_len_ptr,
+                2 as symcrypt_sys::UINT32,
+                NumberFormat::MSB.to_num_format(),
+                key_usage.to_flag(),
+                rsa_key,
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(RsaKeyPair {
+                    inner: rsa_key,
+                    key_usage: key_usage,
+                }),
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    /// `size_of_primes()` returns a tuple containing the sizes, in bytes, of byte arrays large enough to store each of the two primes of the RSA key.
+    pub fn size_of_primes(&self) -> (u32, u32) {
+        unsafe {
+            // SAFETY: FFI calls
+            // Currently, only two prime RSA is supported, i.e. the only valid indexes are 0 and 1
+            let prime_1 = symcrypt_sys::SymCryptRsakeySizeofPrime(self.inner, 0);
+            let prime_2 = symcrypt_sys::SymCryptRsakeySizeofPrime(self.inner, 1);
+            (prime_1, prime_2)
+        }
+    }
+
+    /// `export_key_pair_blob()` returns a [`RsaKeyPairBlob`] value.
+    pub fn export_key_pair_blob(&self) -> Result<RsaKeyPairBlob, SymCryptError> {
+        // Get size of primes only once
+        let (size_p, size_q) = self.size_of_primes();
+
+        // Allocate buffers for filling RsaKeyPairBlob
+        let mut modulus_buffer = vec![0u8; self.size_of_modulus() as usize];
+        let mut pub_exp = vec![0u64; 1];
+        let mut p = vec![0u8; size_p as usize];
+        let mut q = vec![0u8; size_q as usize];
+
+        let mut d_p = vec![0u8; size_p as usize];
+        let mut d_q = vec![0u8; size_q as usize];
+
+        let mut crt_coefficient = vec![0u8; size_p as usize];
+        let mut private_exponent = vec![0u8; self.size_of_modulus() as usize];
+
+        let mut primes_len = [
+            p.len() as symcrypt_sys::SIZE_T,
+            q.len() as symcrypt_sys::SIZE_T,
+        ];
+        let mut crt_lens = [
+            d_p.len() as symcrypt_sys::SIZE_T,
+            d_q.len() as symcrypt_sys::SIZE_T,
+        ];
+
+        unsafe {
+            // SAFETY: FFI calls
+            let result = symcrypt_sys::SymCryptRsakeyGetValue(
+                self.inner,
+                modulus_buffer.as_mut_ptr(),
+                modulus_buffer.len() as symcrypt_sys::SIZE_T,
+                pub_exp.as_mut_ptr(),
+                1 as symcrypt_sys::UINT32,
+                [p.as_mut_ptr(), q.as_mut_ptr()].as_mut_ptr(),
+                primes_len.as_mut_ptr(),
+                2 as symcrypt_sys::UINT32,
+                NumberFormat::MSB.to_num_format(),
+                self.key_usage.to_flag(),
+            );
+            if result != symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR {
+                return Err(result.into());
+            }
+
+            let result = symcrypt_sys::SymCryptRsakeyGetCrtValue(
```

Comment:
>SymCryptRsakeyGetCrtValue [](http://example.com/codeflow?start=39&length=25)

Don't. Just don't. 

These values are non-standardized and the only reason SymCrypt can export them is for backward compatibility with older software that you don't have to worry about. Export only N, e, P, and Q.

Details: the private exponent 'd' is sometimes computed as d = 1/e mod (P-1)(Q-1) and sometimes as d = 1/e mod LCM(P-1,Q-1). The former being more common, the latter being more efficient and more correct. And in some cases the implementation doesn't compute 'd' at all, as only d_p and d_q are necessary.

The CRT coefficients that you need depend on the method the code uses to convert from CRT representation to normal integer representation. There are several methods you can use, and they use different CRT coefficients. 

These values should not be exported or made visible as they only create trouble.

Reply:
Based on a discussion I had with @PhilHallin and this [RFC](https://datatracker.ietf.org/doc/html/rfc8017#appendix-A.1.2), I included the Crt values.

> an RSA private key should be represented with the ASN.1 type
>    RSAPrivateKey:
> 
>          RSAPrivateKey ::= SEQUENCE {
>              version           Version,
>              modulus           INTEGER,  -- n
>              publicExponent    INTEGER,  -- e
>              privateExponent   INTEGER,  -- d
>              prime1            INTEGER,  -- p
>              prime2            INTEGER,  -- q
>              exponent1         INTEGER,  -- d mod (p-1)
>              exponent2         INTEGER,  -- d mod (q-1)
>              coefficient       INTEGER,  -- (inverse of q) mod p
>              otherPrimeInfos   OtherPrimeInfos OPTIONAL
>          }

Reply:
They aren't optional in the above ASN.1 definition.